client/database.go

```diff
@@ -17,107 +17,212 @@ package client
+	return &Database{
+		name:    name,
+		driver:  driver,
+		schemas: make(map[string]*schema.Schema),
+		created: make(map[string]bool),
 	}
 }
 
-func (db *database) Transact(
-	ctx context.Context,
-	fn TxFunc,
-) (interface{}, error) {
-	tx, err := db.driver.BeginTx(ctx, db.name, nil)
-	if err != nil {
-		return nil, fmt.Errorf("error beginning transaction: %w", err)
-	}
-
-	res, err := fn(ctx, tx)
-	if err != nil {
-		if rollbackErr := tx.Rollback(ctx); rollbackErr != nil {
-			return nil, fmt.Errorf(
-				"error trying to rollback transaction: %v, original error: %w",
-				rollbackErr, err)
-		}
-		return nil, fmt.Errorf("error running transaction: %w", err)
-	}
-
-	return res, nil
-}
-
-func (db *database) CreateCollectionsFromModels(ctx context.Context, dbName string, model interface{}, models ...interface{}) error {
-	schemas, err := schema.FromCollectionModels(model, models)
+// CreateCollections creates collections in the Database using provided collection models
+func (db *Database) CreateCollections(ctx context.Context, model schema.Model, models ...schema.Model) error {
+	schemas, err := schema.FromCollectionModels(model, models...)
 	if err != nil {
 		return fmt.Errorf("error parsing model schema: %w", err)
 	}
 
-	return db.CreateCollections(ctx, dbName, schemas)
+	db.schemas = schemas
+
+	return db.createCollectionsFromSchemas(ctx, db.name, schemas)
 }
 
-func (db *database) CreateCollectionsFromDatabaseModel(ctx context.Context, dbModel interface{}) error {
+func (db *Database) createCollectionsFromDatabaseModel(ctx context.Context, dbModel schema.Model) error {
 	dbName, schemas, err := schema.FromDatabaseModel(dbModel)
 	if err != nil {
 		return err
 	}
 
-	return db.CreateCollections(ctx, dbName, schemas)
+	db.schemas = schemas
+
+	return db.createCollectionsFromSchemas(ctx, dbName, schemas)
 }
 
-func (db *database) CreateCollections(ctx context.Context, dbName string, schemas []*schema.Schema) error {
+// createCollectionsFromSchemas transactionally creates collections from the provided schema map
+func (db *Database) createCollectionsFromSchemas(ctx context.Context, dbName string, schemas map[string]*schema.Schema) error {
 	tx, err := db.driver.BeginTx(ctx, dbName)
 	if err != nil {
 		return err
 	}
 	defer func() { _ = tx.Rollback(ctx) }()
 
+	created := make(map[string]bool)
+
 	for _, v := range schemas {
 		sch, err := schema.Build(v)
 		if err != nil {
 			return err
 		}
-		log.Debug().Interface("schema", v).Str("collection", v.Name).Msg("migrateModel")
 		err = tx.CreateOrUpdateCollection(ctx, v.Name, sch)
 		if err != nil {
 			return err
 		}
+		created[v.Name] = true
+	}
+
+	if err := tx.Commit(ctx); err != nil {
+		return err
+	}
+
+	db.created = created
+
+	return nil
+}
+
+// GetCollection returns collection object corresponding to collection model T
+func GetCollection[T schema.Model](ctx context.Context, db *Database) (*Collection[T], error) {
```

Comment:
What happens if GetCollection is called but the database is not created yet? I mean is there some easy way for the user to know that first CreateDatabase needs to be called on the db object?

Reply:
It will fail with "database doesn't exist"